### ISSUE TITLE
Styling console/terminal output

### ISSUE THREAD

**Issue by scls19fr** [pandas-dev/pandas#18066]
Hello,

Since v0.17.1, Pandas have an interesting API for styling Jupyter HTML outputs for Series / DataFrames...
https://pandas.pydata.org/pandas-docs/stable/style.html

It will be a nice feature to be able to have styling for console/terminal output
I wonder if such feature should be part of Pandas core or be inside an other package (like [tabulate](https://bitbucket.org/astanin/python-tabulate) from @astanin which was mentioned in #11052

There is also [PrettyPandas](https://github.com/HHammond/PrettyPandas) from @HHammond

Kind regards

**Comment by gfyoung** [comment]
@scls19fr : Thanks for the report!  I think in the interest of modularity, we would probably want to move away from incorporating giant chunks of code into our codebase.  That being said, we *could* add functions that *call* those function from something like `tabulate` or `PrettyPandas` (assuming they're installed).  However, we would probably need to choose one or the other (or make a uniform API to be able to call both).

**Comment by cbrnr** [comment]
RStudio adds color to the output of a data frame in the console:

![dxm8a7uw4aakwvg](https://user-images.githubusercontent.com/4377312/37096711-ccb6e1ca-2219-11e8-80e1-27f3954bacc0.jpg)

Since IPython now supports syntax highlighting in a normal console, I was wondering if something similar was possible in pandas. This would likely not require any additional packages, but instead adding color codes to the `__str__` or `__repr__` methods could be all there needs to be done.

**Comment by jreback** [comment]
yes this certainly could be done. would likely take a separate formatting method (just to avoid having string / html / color strings code) in one place.

**Comment by cbrnr** [comment]
Great! What do you mean with separate method? Ideally, color strings should be used when I type `df` in IPython, so this has to be in `__repr__`, right?

**Comment by cbrnr** [comment]
ANSI escape color codes work, so it is pretty straightforward to get colorized output, e.g.:

![screen shot 2018-03-27 at 10 45 05](https://user-images.githubusercontent.com/4377312/37956699-59d2828e-31ac-11e8-83e0-3a086e8a284c.png)

The bigger questions are:

* Where do we want to add this functionality? Currently, I'm modifying the string representation in `pandas.io.formats.format.to_string` (before `strcols` get combined into `text`). 
* What kind of implementation would be suitable? Do we add a helper function (e.g. `colorize`)?
* What are the defaults and how do you change them? E.g. there are white on black terminals, black on white terminals, and everything in between.

**Comment by cbrnr** [comment]
I guess it's doable in `to_string`, but carrying around defaults and arguments might require some thinking. I wouldn't call it minefield, but the way the final string is constructed forced me to apply colors in the previous step (in the list of str_cols).

If there is an external tool that can do this coloring for us, we should use it if it means less work implementing it. It could be an optional dependency, in the sense that if it's not installed there will only be plain non-colored output.

**Comment by rswgnu** [comment]
@cbrnr wrote:
> ANSI escape color codes work, so it is pretty straightforward to get colorized output

Say I wanted to colorize just specific table cells upon output to highlight them, based on a list of iloc-compatible cell locations.  How would I do that?  I just need some pointers on what Pandas functions to tweak to allow this.  Thanks.

**Comment by texxronn** [comment]
I could do somehting like

`from colorama import Fore, Back, Style`

`df[c] = Fore.RED + Style.BRIGHT + df[c].astype(str) + Style.RESET_ALL`

print (df)

**Comment by cbrnr** [comment]
Interesting. This was already suggested in #459, but never implemented.

**Comment by rswgnu** [comment]
Thanks for pointing me to this.  It would be nice to have this as a core Pandas capability but at least terminal styling is pissible with this package.

-- Bob

**Comment by ghost711** [comment]
I was looking for a way to use ANSI color codes in the terminal or qtconsole for a long time, and finally put this together. 

It's too hacky for a pull request, but it solves my problem, so I thought I'd post it in case it helps anyone else.

You can replace the "TextAdjustment" class with the version below in this file:
lib/python3.6/site-packages/pandas/io/formats/format.py

`
class TextAdjustment(object):  
    def __init__(self):
        import re
        self.ansi_regx = re.compile(r'\x1B[@-_][0-?]*[ -/]*[@-~]')
        self.encoding = get_option("display.encoding")
    
    def len(self, text): 
        return len( self.ansi_regx.sub('', text) ) 
            
    def justify(self, texts, max_len, mode='right'):         
        unescaped = [self.ansi_regx.sub('', s) for s in texts];     
        escapes = [self.ansi_regx.findall(t) if (len(ue) > 0) else ['', ''] 
                                         for t, ue in zip(texts, unescaped)]  
        if   mode == 'left':   
            js = [s1.ljust( max_len) for s1 in unescaped]
        elif mode == 'center':
            js = [s1.center(max_len) for s1 in unescaped]
        else:                  
            js = [s1.rjust( max_len) for s1 in unescaped]  
        return [((e[0].strip() + j + e[1].strip()) if (len(e) > 0) else j) 
                                                for j, e in zip(js, escapes)]   
    
    def adjoin(self, space, *lists, **kwargs):
        return adjoin(space, *lists, strlen=self.len, justfunc=self.justify, **kwargs)
`

**Comment by Beanking77** [comment]
I got similar requirements to colorize specific column and here's my workaround:
```
from colorama import Fore, Back, Style

def color_red_green(val):
    if val < 0:
        color = Fore.GREEN
    else:
        color = Fore.RED
    return color + str('{0:.2%}'.format(val)) + Style.RESET_ALL

# apply to specific column
dfs["percent"] = dfs["percent"].apply(color_red_green)
```
thanks @texxronn

**Comment by cscanlin-kwh** [comment]
Trying to use corlorama on a pandas dataframe, but running into the same problem with misaligned printing menioned above: https://github.com/pandas-dev/pandas/issues/18066#issuecomment-522192922

Does anyone know of a way to patch this behavior in without modifying pandas itself?

I see [pandas.DataFrame.to_string](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_string.html) has a `formatters` parameter, but it's not clear to me how to use it.

**Comment by TomAugspurger** [comment]
@cscanlin I think pandas will need to be updated to handle this. https://github.com/pandas-dev/pandas/pull/30778 had a start, which allows things like


![](https://user-images.githubusercontent.com/1312546/71818268-179d7980-304e-11ea-88ae-3ead88c8ea52.png)

I'm not planning to return to that anytime soon, so feel free to take over if you want.

**Comment by Erotemic** [comment]
I've updated the workaround from @ghost711 to work with pandas 1.2.4. Hopefully this feature will be supported in the main branch somehow, but in the meantime: 

```python
def monkeypatch_pandas():
    """
    References:
        https://github.com/pandas-dev/pandas/issues/18066
    """
    import pandas.io.formats.format as format_
    from six import text_type

    # Made wrt pd.__version__ == '1.2.4'

    class TextAdjustmentMonkey(object):
        def __init__(self):
            import re
            self.ansi_regx = re.compile(r'\x1B[@-_][0-?]*[ -/]*[@-~]')
            self.encoding  = format_.get_option("display.encoding")

        def len(self, text):
            return len(self.ansi_regx.sub('', text))

        def justify(self, texts, max_len, mode='right'):
            jfunc = str.ljust if (mode == 'left')  else \
                    str.rjust if (mode == 'right') else str.center
            out = []
            for s in texts:
                escapes = self.ansi_regx.findall(s)
                if len(escapes) == 2:
                    out.append(escapes[0].strip() +
                               jfunc(self.ansi_regx.sub('', s), max_len) +
                               escapes[1].strip())
                else:
                    out.append(jfunc(s, max_len))
            return out

        def _join_unicode(self, lines, sep=''):
            try:
                return sep.join(lines)
            except UnicodeDecodeError:
                sep = text_type(sep)
                return sep.join([x.decode('utf-8') if isinstance(x, str) else x
                                                                for x in lines])

        def adjoin(self, space, *lists, **kwargs):
            # Add space for all but the last column:
            pads = ([space] * (len(lists) - 1)) + [0]
            max_col_len = max([len(col) for col in lists])
            new_cols = []
            for col, pad in zip(lists, pads):
                width = max([self.len(s) for s in col]) + pad
                c     = self.justify(col, width, mode='left')
                # Add blank cells to end of col if needed for different col lens:
                if len(col) < max_col_len:
                    c.extend([' ' * width] * (max_col_len - len(col)))
                new_cols.append(c)

            rows = [self._join_unicode(row_tup) for row_tup in zip(*new_cols)]
            return self._join_unicode(rows, sep='\n')

    format_.TextAdjustment = TextAdjustmentMonkey
```

**Comment by brechtm** [comment]
For what it's worth, I found it pretty frustrating that even something basic like representing NaNs differently on the terminal is impossible. For HTML and LaTeX output, there's the Styler and its _na_rep_ argument, but there doesn't seem to be something equivalent for terminal output.

I'm currently working around this by replacing NaNs with `math.inf` (using _fillna_) and providing custom _float_format_ function:

```python
def float_format(value: float):
    if isinf(value):
        return ''
    return str(value)

pd.options.display.float_format = float_format
```

A simple solution would be to also pass NaNs through the _float_format_ function, but I guess that's not an option considering backward compatibilty.

**Comment by cbrnr** [comment]
I still think that rendering the output with Rich would be the best option. Basically, `pd.DataFrames` could have a `__rich_repr__()` method, which is responsible for the repr when Rich is available (https://rich.readthedocs.io/en/stable/pretty.html). Unfortunately, I don't have time to implement it myself at the moment.
